packages/material-ui/src/ListItem/ListItem.js

```diff
@@ -1,18 +1,63 @@
 import * as React from 'react';
 import PropTypes from 'prop-types';
 import clsx from 'clsx';
-import { chainPropTypes, elementTypeAcceptingRef } from '@material-ui/utils';
-import withStyles from '../styles/withStyles';
+import { unstable_composeClasses as composeClasses } from '@material-ui/unstyled';
+import { deepmerge, chainPropTypes, elementTypeAcceptingRef } from '@material-ui/utils';
+import experimentalStyled from '../styles/experimentalStyled';
 import { alpha } from '../styles/colorManipulator';
 import ButtonBase from '../ButtonBase';
 import isMuiElement from '../utils/isMuiElement';
 import useEnhancedEffect from '../utils/useEnhancedEffect';
 import useForkRef from '../utils/useForkRef';
 import ListContext from '../List/ListContext';
+import listItemClasses, { getListItemUtilityClass } from './listItemClasses';
 
-export const styles = (theme) => ({
-  /* Styles applied to the (normally root) `component` element. May be wrapped by a `container`. */
-  root: {
+const overridesResolver = (props, styles) => {
+  const { styleProps } = props;
+
+  return deepmerge(styles.root || {}, {
+    ...(styleProps.dense && { dense: styles.dense }),
+    ...(styleProps.alignItems === 'flex-start' && { alignItemsFlexStart: styles.alignItemsFlexStart }),
+    ...(styleProps.divider && { divider: styles.divider }),
+    ...(!styleProps.disableGutters && { gutters: styles.gutters }),
+    ...(styleProps.button && { button: styles.button }),
+    ...(styleProps.hasSecondaryAction && { secondaryAction: styles.secondaryAction }),
+    ...(styleProps.selected && { selected: styles.selected }),
+    ...(styleProps.disabled && { disabled: styles.disabled }),
+  });
+};
+
+const useUtilityClasses = (styleProps) => {
+  const { dense, alignItems, divider, disableGutters, hasSecondaryAction, selected, disabled, button, classes } = styleProps;
+
+  const slots = {
+    root: [
+      'root',
+      dense && 'dense',
+      alignItems === 'flex-start' && 'alignItemsFlexStart',
+      divider && 'divider',
+      !disableGutters && 'gutters',
+      button && 'button',
+      hasSecondaryAction && 'secondaryAction',
+      selected && 'selected',
+      disabled && 'disabled',
+    ],
+    container: ['container'],
+  };
+
+  return composeClasses(slots, getListItemUtilityClass, classes);
+};
+
+const getListItemRoot = component => (experimentalStyled(
```

Comment:
```suggestion
const ListItemRoot = experimentalStyled(
```
We should never use methods for generating components, as it would create new one on each invocation. You can use the `as` prop to provide different component.